crates/solidity/outputs/cargo/crate/src/backend/ir/ast/ast_nodes.rs.jinja2

```diff
@@ -0,0 +1,192 @@
+          pub fn {{ field.label }}(&self) -> Option<{{ field.type.name }}> {
+            self.ir_node.{{ field.label }}.as_ref().map(|ir_node| {
+              Rc::new({{ field.type.name }}Struct::create(ir_node, &self.semantic))
+            })
+          }
+        {% elif field.type.kind == "UniqueTerminal" %}
+          pub fn {{ field.label }}(&self) -> bool {
+            self.ir_node.{{ field.label }}
+          }
+        {% elif field.type.is_terminal %}
+          pub fn {{ field.label }}(&self) -> Option<Rc<TerminalNode>> {
+            self.ir_node.{{ field.label }}.as_ref().map(Rc::clone)
+          }
+        {% elif target.collections is containing(field.type.name) %}
+          {% set item_type = target.collections[field.type.name].item_type %}
+          {% if custom_types is containing(item_type.name) or not item_type.is_terminal %}
+            pub fn {{ field.label }}(&self) -> Option<{{ field.type.name }}> {
+              self.ir_node.{{ field.label }}.as_ref().map(|ir_node| {
+                Rc::new({{ field.type.name }}Struct::create(ir_node, &self.semantic))
+              })
+            }
+          {% elif item_type.is_terminal %}
+            pub fn {{ field.label }}(&self) -> Option<Vec<Rc<TerminalNode>>> {
+              self.ir_node.{{ field.label }}.as_ref().map(|items| items.clone())
+            }
+          {% endif -%}
+        {% endif %}
+      {% else %}
+        {% if custom_types is containing(field.type.name) or
+              target.sequences is containing(field.type.name) or
+              target.choices is containing(field.type.name) %}
+          pub fn {{ field.label }}(&self) -> {{ field.type.name }} {
+            Rc::new({{ field.type.name }}Struct::create(&self.ir_node.{{ field.label }}, &self.semantic))
+          }
+        {% elif field.type.is_terminal %}
+          pub fn {{ field.label }}(&self) -> Rc<TerminalNode> {
+            Rc::clone(&self.ir_node.{{ field.label }})
+          }
+        {% elif target.collections is containing(field.type.name) %}
+          {% set item_type = target.collections[field.type.name].item_type %}
+          {% if custom_types is containing(item_type.name) or not item_type.is_terminal %}
+            pub fn {{ field.label }}(&self) -> {{ field.type.name }} {
+              Rc::new({{ field.type.name }}Struct::create(&self.ir_node.{{ field.label }}, &self.semantic))
+            }
+          {% elif item_type.is_terminal %}
+            pub fn {{ field.label }}(&self) -> Vec<Rc<TerminalNode>> {
+              self.ir_node.{{ field.label }}.clone()
+            }
+          {% endif %}
+        {% endif %}
+      {% endif %}
+    {% endfor %}
+  }
+
+{% endfor %}
+
+//
+// Choices:
+//
+
+{% for parent_type, choice in target.choices %}
+  pub type {{ parent_type }} = Rc<{{ parent_type }}Struct>;
+
+  pub struct {{ parent_type }}Struct {
+    pub(crate) ir_node: input_ir::{{ parent_type }},
+    semantic: Rc<SemanticAnalysis>,
+  }
```

Comment:
I wonder if this could be represented with an `enum`, it feels we're going against the language (Rust) by representing it with a struct.

Maybe keeping this struct, but then having a single function that returns an enum with one of the options.

It could also be faster